### ISSUE TITLE
Make wpt own the HTML parser test data and remove dependency on html5lib-python, html5lib-tests

### ISSUE THREAD

**Issue by zcorpan** [web-platform-tests/wpt#27868]
This week I've done the exercise of updating HTML parser tests again, though this time I was a bit more successful in figuring out how to get those changes through to wpt (see #2887). But boy is it painful and also mostly undocumented!

* Make the test data change to `html5lib-tests` (in the custom test data format) https://github.com/html5lib/html5lib-tests/pull/133
* Update `html5lib-python`'s submodule of `html5lib-tests` AND update `.pytest.expect` (manually?) so that html5lib itself doesn't fail the changed tests without having them marked as expected failures. https://github.com/html5lib/html5lib-python/pull/531
* Update the commit hash for `html5lib-python` in wpt's `html/tools/build.sh` and generate tests in `wpt` by running `html/tools/build.sh`. https://github.com/web-platform-tests/wpt/pull/27799

Juggling 3 repos for one change like this doesn't seem ideal for contributors. From wpt's perspective, what I would like instead is:

* Make the test data change in `wpt` and run a script to generate tests by running a script. No dependency on html5lib.

Then `html5lib-python` can get the tree-builder test data from `wpt` instead of from `html5lib-tests`.

Thoughts? @gsnedders @jgraham @annevk @stephenmcgruer

**Comment by gsnedders** [comment]
this is effectively a dupe of https://github.com/html5lib/html5lib-tests/issues/127 fwiw

**Comment by zcorpan** [comment]
@gsnedders oh, right, I had forgotten about that! It seems like there isn't objection. Are you still planning to work on this?

**Comment by gsnedders** [comment]
> @gsnedders oh, right, I had forgotten about that! It seems like there isn't objection. Are you still planning to work on this?

It is a long way down my list.

**Comment by zcorpan** [comment]
A tweak we can make is to depend on `html5lib-tests` instead of `html5lib-python` from `wpt`, which would remove the second step. (I think this was @jgraham 's idea, but don't see it mentioned in GitHub.)

**Comment by gsnedders** [comment]
One obvious (easy) tweak given it's using `git-submodules` is to explicitly store a commit hash somewhere in WPT and then during update `cd html5lib-python/html5lib/tests/testdata && git fetch origin && git checkout $REV`.

**Comment by hsivonen** [comment]
My main concern is that I want to preserve the file format for the preferred form form making modifications to the test, since there are non-WPT consumers of those formats.

I'm not a fan of WPT having a build step that transforms the tree builder test format. FWIW, Gecko's mochitest harness stores the original .dat format in the repo and [parses it when the tests are run](https://searchfox.org/mozilla-central/source/parser/htmlparser/tests/mochitest/parser_datreader.js).

**Comment by annevk** [comment]
Having worked on a parser bug in WebKit I now think this would be even more valuable than I previously thought. It looks like Chromium and WebKit both have two sets of parser tests in the tree:

* Some html5lib-tests fork of unspecified vintage
* web-platform-tests's import of html5lib-tests

And the former has tests the latter might not contain. I contributed further to this problem in https://github.com/WebKit/WebKit/pull/12019, but am willing to be part of the cleanup crew if we make web-platform-tests the true home of HTML parser tests.

I suspect @mfreed7 might be interested in this from the Chromium side. Copying here to gather interest.

**Comment by mfreed7** [comment]
I'm definitely supportive of the effort to clean this up, and make WPT the source of truth for parser tests.

**Comment by annevk** [comment]
Steps taken thus far:

* Upstreamed WebKit-specific tests: https://github.com/html5lib/html5lib-tests/commit/4f45c0211cf1d1f1af319470f77851f60f29914c
* Working on a new import in https://github.com/web-platform-tests/wpt/pull/39305

I wonder if @zcorpan is still interested in taking this even further as I think it would definitely be preferable if we didn't have to go via html5lib-tests.

https://github.com/html5lib/html5lib-tests does have a number of actionable issues and stale PRs worth triaging. Help appreciated.

**Comment by zcorpan** [comment]
Yes. See https://github.com/html5lib/html5lib-tests/issues/127#issuecomment-1490501826 and later comments.

**Comment by annevk** [comment]
@zcorpan any progress on this?

**Comment by zcorpan** [comment]
Not yet but it's on my list.

**Comment by Anutrix** [comment]
Friendly bump-up.

`html5lib-python` seems pretty much dead. Last commit was Feb 2024. Even removal of six and other PRs are open for over a year now. It's time we look for alternatives.

**Comment by annevk** [comment]
I've now imported the `.dat` files, created https://github.com/web-platform-tests/wpt/pull/60313 for a final bump, and posted an update to https://github.com/html5lib/html5lib-tests/issues/127#issuecomment-4592630737.

I suggest that after we land that referenced PR here we stop being downstream from html5lib-tests.

**Comment by gsnedders** [comment]
@annevk To check my understanding: we don’t have the history of the tests in this repo, and we no longer have linting for the data file format (which had been introduced previously after people messed it up)?

**Comment by annevk** [comment]
That seems correct, but we should be able to add linting. Not sure test history is that important here.

**Comment by josepharhar** [comment]
Thanks for working on this! How should we go about making parser changes in the future? Update them in WPT and html5lib separately? Only update WPT and ignore html5lib? Something else?

**Comment by gsnedders** [comment]
> Not sure test history is that important here.

Knowing that a test was added due to a web compatibility bug, for example, is very much useful information — it suggests that that behaviour goes beyond pure spec correctness.

**Comment by annevk** [comment]
@josepharhar only WPT.
